In [2]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
#tutorial https://www.kaggle.com/code/prashant111/random-forest-classifier-tutorial
data = "./features-relative.csv"
df = pd.read_csv(data)
patient_id = "BA0803901"

df_test = df[df['file'].str.contains(patient_id)]

# Train set: all other patients
df_train = df[~df['file'].str.contains(patient_id)]

X = df.drop(['Label', 'Start', 'End', 'file', 'Start Sample', 'End Sample'], axis=1)
X_train = df_train.drop(['Label', 'Start', 'End', 'file', 'Start Sample', 'End Sample'], axis=1)
X_test = df_test.drop(['Label', 'Start', 'End', 'file', 'Start Sample', 'End Sample'], axis=1)

# Labels
y_train = df_train['Label']
y_test = df_test['Label']

# Encode labels
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
y_train = le.fit_transform(y_train)
y_test = le.transform(y_test)  # only transform, don't fit again

KeyError: "['Label', 'End', 'Start Sample', 'End Sample'] not found in axis"

In [ ]:

from sklearn.ensemble import RandomForestClassifier


rfc = RandomForestClassifier(random_state=0, n_estimators=300, class_weight="balanced")

rfc.fit(X_train, y_train)

y_pred = rfc.predict(X_test)

from sklearn.metrics import accuracy_score

print('Model accuracy score with 100 decision-trees : {0:0.4f}'. format(accuracy_score(y_test, y_pred)))

Model accuracy score with 100 decision-trees : 0.7799


In [ ]:
from sklearn.model_selection import RandomizedSearchCV

param_grid = {
    'n_estimators':[200,400,600],
    'max_depth':[None, 20, 30, 40],
    'min_samples_split':[2,5,10],
    'min_samples_leaf':[1,2,4],
    'max_features':['sqrt','log2']
}

rfc = RandomForestClassifier(random_state=42)
search = RandomizedSearchCV(rfc, param_distributions=param_grid, n_iter=50, cv=5, scoring='accuracy', n_jobs=-1)
search.fit(X_train, y_train)
print(search.best_params_)


KeyboardInterrupt: 

In [ ]:
feature_scores = pd.Series(rfc.feature_importances_, index=X_train.columns).sort_values(ascending=False)

print(feature_scores)

spec_entropy-3     0.024298
spec_entropy-10    0.024197
spec_entropy-1     0.023980
spec_entropy-2     0.023728
spec_entropy-5     0.022759
                     ...   
zcr-4              0.000483
zcr-8              0.000459
zcr-10             0.000409
zcr-7              0.000392
zcr-9              0.000272
Length: 120, dtype: float64

In [ ]:
print(df['file'].value_counts())



file
.\Spontanaktivität\filtered_BP0803902_segment_1.wav     544
.\Spontanaktivität\filtered_BA0803901_segment_3.wav     423
.\Spontanaktivität\filtered_LI06056410_segment_1.wav    403
.\Spontanaktivität\filtered_HC280963_segment_1.wav      369
.\Spontanaktivität\filtered_RN181281_segment_3.wav      318
.\Spontanaktivität\filtered_BA0803901_segment_2.wav     247
.\Spontanaktivität\filtered_LI0605644_segment_1.wav     247
.\Spontanaktivität\filtered_HC280963_segment_2.wav      237
.\Spontanaktivität\filtered_LI0605645_segment_1.wav     194
.\Spontanaktivität\filtered_LI0605645_segment_4.wav     192
.\Spontanaktivität\filtered_LI0605645_segment_3.wav     175
.\Spontanaktivität\filtered_QD161095_segment_5.wav      170
.\Spontanaktivität\filtered_RN181281_segment_2.wav      167
.\Spontanaktivität\filtered_LI0605648_segment_2.wav     162
.\Spontanaktivität\filtered_BA0803901_segment_1.wav     157
.\Spontanaktivität\filtered_RN181281_segment_1.wav      156
.\Spontanaktivität\filtered_LI06056

Accuracy with normal split 0.89 accuracy
with patient split 0.78